### PR TITLE
LF: exhaustive tests for CommandPreprocessor

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
While implementing the tests I found a minor bug on preprocessing of Fetch command.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala [hunk 1]
```diff
@@ -51,13 +71,120 @@ class CommandPreprocessorSpec extends AnyWordSpec with Matchers with TableDriven
...
+          a[Error.Preprocessing.Lookup],
+        validFetchByKey.copy(key = ValueList(FrontStack(ValueInt64(42)))) ->
+          a[Error.Preprocessing.TypeMismatch],
+        validLookup.copy(templateId = "Mod:Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validLookup.copy(contractKey = ValueList(FrontStack(ValueInt64(42)))) ->
+          a[Error.Preprocessing.TypeMismatch],
+      )
+
+      forEvery(negativeTestCases) { command =>
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala:171]: I’m confused why is negative = success? I would have expected the two to be flipped.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala [hunk 2]
```diff
@@ -51,13 +71,120 @@ class CommandPreprocessorSpec extends AnyWordSpec with Matchers with TableDriven
...
+        Try(defaultPreprocesseor.unsafePreprocessCommand(command)) shouldBe a[Success[_]]
+      }
+
+      forEvery(positiveTestCases) { (command, typ) =>
+        inside(Try(defaultPreprocesseor.unsafePreprocessCommand(command))) {
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala:176]:
```suggestion
        inside(Try(defaultPreprocessor.unsafePreprocessCommand(command))) {
```

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala [hunk 3]
```diff
@@ -51,13 +71,120 @@ class CommandPreprocessorSpec extends AnyWordSpec with Matchers with TableDriven
...
+          a[Error.Preprocessing.Lookup],
+        validExe.copy(choiceId = "Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validExe.copy(argument = ValueRecord("", ImmArray("content" -> ValueInt64(42)))) ->
+          a[Error.Preprocessing.TypeMismatch],
+        validExeByKey.copy(templateId = "Mod:Undefined") ->
+          a[Error.Preprocessing.Lookup],
+        validExeByKey.copy(contractKey = ValueList(FrontStack(ValueInt64(42)))) ->
+          a[Error.Preprocessing.TypeMismatch],
+        validExeByKey.copy(choiceId = "Undefined") ->
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala:143]:
```suggestion
        validExeByKey.copy(contractKey = ValueList(FrontStack(ValueInt64(42)))) ->
          a[Error.Preprocessing.TypeMismatch],
        validExeByKey.copy(choiceId = "Undefined") ->
```
nit: maybe consider swapping these two to be consistent with the other cases that first handle lookups and then type mismatches.

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala]: I renamed to `noErrorTestCases` /`errorTestCases`

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala (hunk 3 above)
Comment (cocreature) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/CommandPreprocessorSpec.scala:143]: ah makes sense, didn’t realize it was in field order :+1:

**Pull request merged**